interfaces/apparmor/apparmor.go

```diff
@@ -0,0 +1,94 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package apparmor contains primitives for working with apparmor.
+//
+// References:
+//  - http://wiki.apparmor.net/index.php/Kernel_interfaces
+//  - http://apparmor.wiki.kernel.org/
+//  - http://manpages.ubuntu.com/manpages/xenial/en/man7/apparmor.7.html
+package apparmor
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// LoadProfile loads an apparmor profile from the given file.
+//
+// If no such profile was previously loaded then it is simply added to the kernel.
+// If there was a profile with the same name before, that profile is replaced.
+func LoadProfile(fname string) error {
+	return exec.Command("apparmor_parser", "--replace", fname).Run()
+}
+
+// Profile contains the name and mode of an apparmor profile loaded into the kernel.
+type Profile struct {
+	// Name of the profile. This is is either full path of the executable or an
+	// arbitrary string without spaces.
+	Name string
+	// Mode is either "enforcing" or "complaining".
+	Mode string
+}
+
+// Unload removes a profile from the running kernel.
+//
+// The operation is done with: apparmor_parser --remove $name
+func (profile *Profile) Unload() error {
+	return exec.Command("apparmor_parser", "--remove", profile.Name).Run()
+}
+
+// profilesPath contains information about the currently loaded apparmor profiles.
+const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
+
+var profilesPath = realProfilesPath
+
+// LoadedProfiles interrogates the kernel and returns a list of loaded apparmor profiles.
+//
+// Snappy manages apparmor profiles named *.snap. Other profiles might exist on
```

Comment:
On Fri, 2016-03-11 at 00:45 -0800, Zygmunt Krynicki wrote:
> > 
> > +
> > +// Unload removes a profile from the running kernel.
> > +//
> > +// The operation is done with: apparmor_parser --remove $name
> > +func (profile *Profile) Unload() error {
> > +	return exec.Command("apparmor_parser", "--remove",
> > profile.Name).Run()
> > +}
> > +
> > +// profilesPath contains information about the currently loaded
> > apparmor profiles.
> > +const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
> > +
> > +var profilesPath = realProfilesPath
> > +
> > +// LoadedProfiles interrogates the kernel and returns a list of
> > loaded apparmor profiles.
> > +//
> > +// Snappy manages apparmor profiles named *.snap. Other profiles
> > might exist on
> Use the filesystem in /var/lib/snappy/apparmor/profiles as your
> database, not the kernel. You already are in many senses since on
> reboot it is the filesystem that will have the profiles you load into
> the kernel. Even if you are of the opinion that bootstrapping is
> separate from that, whatever you bootstrap from is your actual
> database
> and not the kernel. You can Ensure() that profiles are loaded in the
> kernel based on the profiles that exist in the filesystem. You can
> Ensure() that a profile is unloaded from the kernel if the file
> exists
> but the profile is not loaded.
> 
> ZK: This makes no sense. Yes we manage the /var/lib/snappy/apparmor/*
> space but I don't see how this is related. My concern is solely about
> profiles loaded into the kernel.
> 
> When the snap manager changes a snap (e.g. updates it to a newer
> version that has different apps inside) we must generate new profiles
> (which is easy) and discard old profiles. Since profiles are a
> runtime property of the kernel we need a way to discover what
> profiles exist, and, to the best of our knowledge, which profiles to
> remove
> 
/var/lib/snappy/apparmor/* is directly related because those (and
/var/cache/apparmor/*) are what are loaded in the kernel to add to the
list that is in /sys/kernel/security/apparmor/profiles. Put another
way: you can always compare what is in the filesystem to what is loaded
in the kernel. Ie, ls /var/lib/snappy/apparmor/* to get the full list
of profiles, then look in /sys/kernel/security/apparmor/profiles and
see what is loaded and unloaded and makes decisions from there. It is
very easy to then see what exists in the kernel that you are
responsible for and which profiles are not in the kernel (that perhaps
should be). The current state then is-- see what in
/var/lib/snappy/apparmor/* is loaded in the kernel.


-- 
Jamie Strandboge             | http://www.canonical.com